### PR TITLE
Fix tests against jupyter_server 2

### LINKED ISSUE #362: Tests broken for jupyter_server 2 - help wanted!
Our jupyterlab extension tests for JupyterLab 2 and JupyterLab 3 are both failing when jupyter_server 2 is used, which it is in our Python 3.11 tests but not in our Python 3.7 tests because jupyter_server 2 requires python 3.8 and is not installed in the python 3.7 context.

# Summary to help understand the error

I've debugged this in a fork where I run tests in a Python 3.8 environment, and only switch between installing jupyter_server 1 and 2. Doing so, I conclude the following two separate errors we encounter when testing with jupyterlab 2 and 3 together with jupyter_server 2.

## JupyterLab 2 failure

When using jupyterlab 2 and jupyter_server 2 in a Python 3.8 environment, we fail in the `python -m jupyterlab.browser_check` step, and here are some diff from logs and python dependencies installed.

```diff
-# lab 2, py3.8, jupyter_server 1, notebook
+# lab 2, py3.8, jupyter_server 2, notebook
    app dir: /opt/hostedtoolcache/Python/3.8.15/x64/share/jupyter/lab
         @jupyterhub/jupyter-server-proxy v3.2.2  enabled  OK*
 JupyterLab v2.3.2
@@ -83,11 +83,21 @@ Starting Chrome Headless
 Navigating to page: file:///home/runner/.local/share/jupyter/runtime/nbserver-x-open.html
 Waiting for page to load...
 Waiting for application to start...
+/opt/hostedtoolcache/Python/3.8.15/x64/lib/python3.8/site-packages/jupyter_server/base/handlers.py:240: UserWarning: The Tornado web application does not have an 'identity_provider' defined in its settings. In future releases of jupyter_server, this will be a required key for all subclasses of `JupyterHandler`. For an example, see the jupyter_server source code for how to add an identity provider to the tornado settings: https://github.com/jupyter-server/jupyter_server/blob/aa8fd8b3faf37466eeb99689d5555314c5bf6640/jupyter_server/serverapp.py#L253
+  warnings.warn(
+[E x BrowserApp] Error unpacking user from cookie: Extra data: line 1 column 2 (char 1)
+Error unpacking user from cookie: Extra data: line 1 column 2 (char 1)
+[W x BrowserApp] Clearing invalid/expired login cookie username-127-0-0-1-8889
+[I x BrowserApp] 302 GET /foo/server-proxy/servers-info (127.0.0.1) 5.060000ms
+Parsed an error from text content: undefined Plugin '@jupyterhub/jupyter-server-proxy:main' failed to activate.
+Parsed an error from text content: undefined SyntaxError: Unexpected token < in JSON at position 0
 Chrome test complete
 [I x BrowserApp] Test Complete
-[I x BrowserApp] Exiting normally
+[C x BrowserApp] Exiting with 1 due to errors
+Exiting with 1 due to errors
 [I x BrowserApp] Stopping server...
 [W x BrowserApp] Failed to fetch package metadata for '@jupyterhub/jupyter-server-proxy': <HTTPError 404: 'Not Found'>
+Error: Process completed with exit code 1.
 
-jupyter-server==1.23.4
+jupyter_server==2.0.6
```

## JupyterLab 3 failure

When using jupyterlab 3 and jupyter_server 2 in a Python 3.8 environment, we fail in the "browser acceptance test" section, and pass the `python -m jupyterlab.browser_check` step that we fail with when using jupyterlab 2.

```diff
-# jupyter_server 1, lab 3, py3.8, notebook
+# jupyter_server 2, lab 3, py3.8, notebook
 ============================= test session starts ==============================
 platform linux -- Python 3.8.15, pytest-7.2.0, pluggy-1.0.0
 rootdir: /home/runner/work/jupyter-server-proxy/jupyter-server-proxy
@@ -12,32 +12,62 @@ Acceptance.Classic :: Server Proxies in Notebook Classic
 ==============================================================================
 Notebook Classic Loads                                                | PASS |
 ------------------------------------------------------------------------------
-Launch Browser Tab                                                    | PASS |
+Launch Browser Tab                                                    | FAIL |
+Element with locator 'css:a[role="menuitem"][href*="foo"]' not found.
 ------------------------------------------------------------------------------
-Launch Another Browser Tab                                            | PASS |
+Launch Another Browser Tab                                            | FAIL |
+Element with locator 'css:a[role="menuitem"][href*="bar"]' not found.
 ------------------------------------------------------------------------------
-Acceptance.Classic :: Server Proxies in Notebook Classic              | PASS |
-3 tests, 3 passed, 0 failed
+Acceptance.Classic :: Server Proxies in Notebook Classic              | FAIL |
+3 tests, 1 passed, 2 failed
 ==============================================================================
 Acceptance.Lab :: Server Proxies in Lab                                       
 ==============================================================================
-Lab Loads                                                             | PASS |
+Lab Loads                                                             | FAIL |
+Parent suite setup failed:
+Element 'xpath://script[contains(@id, 'jupyter-config-data')]' did not appear in 5 seconds.
 ------------------------------------------------------------------------------
-Launch Browser Tab                                                    | PASS |
+Launch Browser Tab                                                    | FAIL |
+Parent suite setup failed:
+Element 'xpath://script[contains(@id, 'jupyter-config-data')]' did not appear in 5 seconds.
 ------------------------------------------------------------------------------
-Launch Lab Tab                                                        | PASS |
+Launch Lab Tab                                                        | FAIL |
+Parent suite setup failed:
+Element 'xpath://script[contains(@id, 'jupyter-config-data')]' did not appear in 5 seconds.
 ------------------------------------------------------------------------------
-Acceptance.Lab :: Server Proxies in Lab                               | PASS |
-3 tests, 3 passed, 0 failed
+Acceptance.Lab :: Server Proxies in Lab                               | FAIL |
+Suite setup failed:
+Element 'xpath://script[contains(@id, 'jupyter-config-data')]' did not appear in 5 seconds.
+
+3 tests, 0 passed, 3 failed
 ==============================================================================
-Acceptance :: Acceptance tests for jupyter-server-proxy               | PASS |
-6 tests, 6 passed, 0 failed
+Acceptance :: Acceptance tests for jupyter-server-proxy               | FAIL |
+6 tests, 1 passed, 5 failed
 ==============================================================================
 Output:  /home/runner/work/jupyter-server-proxy/jupyter-server-proxy/build/robot/output.xml
 Log:     /home/runner/work/jupyter-server-proxy/jupyter-server-proxy/build/robot/log.html
 Report:  /home/runner/work/jupyter-server-proxy/jupyter-server-proxy/build/robot/report.html
-.
+F
+
+=================================== FAILURES ===================================
+__________________________________ test_robot __________________________________
+
+    def test_robot():
+        """run acceptance tests with robotframework"""
+        with_robot = pytest.importorskip("JupyterLibrary")
+    
+        if OUTPUT.exists():
+            shutil.rmtree(OUTPUT)
+    
+        return_code = subprocess.call(
+            ["robot", "--consolecolors=on", f"--outputdir={OUTPUT}", str(HERE)],
+            cwd=str(HERE)
+        )
+    
+>       assert return_code == 0
+E       assert 5 == 0
 
+tests/acceptance/test_acceptance.py:22: AssertionError
 =============================== warnings summary ===============================
 ../../../../../opt/hostedtoolcache/Python/3.8.15/x64/lib/python3.8/site-packages/jupyter_server/base/handlers.py:x
   /opt/hostedtoolcache/Python/3.8.15/x64/lib/python3.8/site-packages/jupyter_server/base/handlers.py:x: DeprecationWarning: Jupyter is migrating its paths to use standard platformdirs
@@ -48,7 +78,9 @@ Report:  /home/runner/work/jupyter-server-proxy/jupyter-server-proxy/build/robot
     from jupyter_core.paths import is_hidden
 
 -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
-================= 1 passed, 34 deselected, 1 warning in 27.61s =================
+=========================== short test summary info ============================
+FAILED tests/acceptance/test_acceptance.py::test_robot - assert 5 == 0
+================= 1 failed, 34 deselected, 1 warning in 43.14s =================
 
-jupyter-server==1.23.4
+jupyter_server==2.0.6
```

### PR DESCRIPTION
**Pull request by bollwyvl**
## References

- fixes #362

## Changes

- [x] adds some explicit detection of `jupyter_server` version info
- [x] sets a bottom pin for `robotframework-jupyterlibrary`

## Notes
- i don't really know what's going on with the `jupyter-notebook` CLI command with `jupyter_server@2`, but `jupyter-server` and `jupyter-lab` seem to work

### CODE REVIEW

**Comment by bollwyvl** [comment]
I'd probably just skip the browser_check, not really sure what value it provides over the actual browser-based acceptance tests.

**Comment by bollwyvl** [comment]
Ok, skipping `browser_check` on py311/lab2 comes up clean without changing the package name.

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -111,7 +112,7 @@ jobs:
       - name: Install Acceptance test dependencies
         run: |
           # the acceptance test requires notebook to run
-          pip install "notebook<7" robotframework-jupyterlibrary
+          pip install "notebook<7" "robotframework-jupyterlibrary>=0.4.2"
```
Comment (bollwyvl) [inline .github/workflows/test.yaml:115]: alternately we could hoist these to the `[test]` extra and install that

**Review by manics — APPROVED** [review]
Seems fine to me! Merge if you're happy

**Comment by manics** [comment]
Oh hang on, the https://github.com/jupyterhub/jupyter-server-proxy/blob/main/.github/workflows/test.yaml workflow isn't running

**Comment by bollwyvl** [comment]
I'm away from keyboard for a couple hours, if it's obvious how to fix the
selector, go ahead!

**Comment by consideRatio** [comment]
Thank you soo much for your work on this @bollwyvl!!!! :heart: :tada:!

@manics this was the issue it seems. I'll quit working for today I think but I'm happy to help resolve this later!

image

**Comment by bollwyvl** [comment]
Think it just needs quotes

**Comment by bollwyvl** [comment]
The web editor liked the`${{}}` form better (as there were already quotes, and it doesn't seem to like `"` inside expressions.

File: tests/acceptance/test_acceptance.py [hunk 2]
```diff
@@ -1,22 +1,47 @@
...
+import os
 import shutil
+import subprocess
+
+import pytest
 
 HERE = Path(__file__).parent
 OUTPUT = HERE.parent.parent / "build/robot"
+SERVER_INFO = None
+LAB_INFO = None
```
Comment (consideRatio) [inline tests/acceptance/test_acceptance.py:11]:
```suggestion
JUPYTER_SERVER_INFO = None
```

File: tests/acceptance/test_acceptance.py [hunk 3]
```diff
@@ -1,22 +1,47 @@
...
+    pass
 
 
 def test_robot():
     """run acceptance tests with robotframework"""
     with_robot = pytest.importorskip("JupyterLibrary")
 
+    env = dict(**os.environ)
+
+    if SERVER_INFO is None:
```
Comment (consideRatio) [inline tests/acceptance/test_acceptance.py:26]:
```suggestion
    # JUPYTER_LIBRARY_* env vars documentation:
    # https://robotframework-jupyterlibrary.readthedocs.io/en/stable/LIMITS.html#notebookapp-vs-serverapp
    if JUPYTER_SERVER_INFO is None:
```

**Review by consideRatio — APPROVED** [review]
Added comments and renamed a constant for readability, but this LGTM!

**Comment by consideRatio** [comment]
I updated the title to clarify that this PR is only adjusting the tests, not the software itself.

I also pushed two commits with:
- some inline comments
- a variable name update (`SERVER_INFO` -> `JUPYTER_SERVER_INFO`)
- fix of an existing reference to `matrix.python` to `matrix.python-version` influencing uploaded artifact names

THANK YOU @bollwyvl for your work on this!!! :heart: :tada: :sunflower:

**Pull request merged**